Issue #11446: Update SuppressionFilterTest to use execute

shubh220922 opened the pull request:
Related to #11446

shubh220922 commented:
```
[INFO] PMD version: 6.50.0
[INFO] PMD Failure: com.puppycrawl.tools.checkstyle.filters.SuppressionFilterTest:163 Rule:JUnitTestsShouldIncludeAssert Priority:3 JUnit tests should include assert() or fail().
[INFO] PMD Failure: com.puppycrawl.tools.checkstyle.filters.SuppressionFilterTest:180 Rule:JUnitTestsShouldIncludeAssert Priority:3 JUnit tests should include assert() or fail().
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  06:40 min
[INFO] Finished at: 2022-11-26T11:05:29+05:30
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.19.0:check (default) on project checkstyle: You have 2 PMD violations. For more details see: C:\Users\shubh\OneDrive\Desktop\checkstyle project\checkstyle\target\pmd.xml -> [Help 1]

```

Why is it saying the below:
``JUnit tests should include assert() or fail().``

Also it is pointing me to some pmd violation in pmd.xml.
When I open pmd.xml I find these two errors for the below line of xml code:

Line of code for which error is coming:
```
<pmd xmlns="http://pmd.sourceforge.net/report/2.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://pmd.sourceforge.net/report/2.0.0 http://pmd.sourceforge.net/report_2_0_0.xsd" version="6.50.0" timestamp="2022-11-26T11:05:29.151">
```
Errors:
```
URI is not registered (Settings | Languages & Frameworks | Schemas and DTDs)
Cannot resolve symbol 'http://pmd.sourceforge.net/report_2_0_0.xsd'
```

Associated image:

![image](https://user-images.githubusercontent.com/90137881/204074383-03255f78-d32f-4139-b9fc-3352cd5e3189.png)

nrmancuso commented:
> JUnit tests should include assert() or fail().

This is from conversion to execute, take a look at previous PRs for this issue to investigate how we handled it.

> URI is not registered

This is local error only, from you not registering the URI of the XML schema and downloading it. You can ignore this, but it is a good idea to read about and understand this concept.

romani commented:
surprisingly we do not have examples how to bypass this problem
all existing `execute` are in form
https://github.com/checkstyle/checkstyle/blob/cd3daa9a2a8a035a077b9fcb7559e8db0dcc1242/src/test/java/com/puppycrawl/tools/checkstyle/filters/SuppressWithNearbyCommentFilterTest.java#L450-L455

We need to exclude this test method from PMD validation.

nrmancuso commented:
@romani after thinking on this, perhaps it is not good to suppress such warnings for the `execute` method in general. We might lose some nuances of tests along the way if we do this. All `verify...` methods contain various assertions in the call stack, while `execute` does not (it only runs Checker` over the supplied files).

It might be better to come up with some assertions regarding the state of the cache file in these tests or something.

shubh220922 commented:
@nrmancuso 



> This is from conversion to execute, take a look at previous PRs for this issue to investigate how we handled it.

I have gone through everyone's PR who had to change method to ``execute()``, seems like I am facing this issue. What's the plan now? Should we exclude this test method from PMD validation? 



> All `verify...` methods contain various assertions in the call stack, while `execute` does not (it only runs Checker` over the supplied files).
> 
> It might be better to come up with some assertions regarding the state of the cache file in these tests or something.

I think it would be good to include some assertions in the call stack of execute. It may solve the problem.

romani commented:
Such test existing only to cover something in cache implementation, not a filter.
Filters ideally should use verifySuppressedWithParser.
Can we use it ?
If yes - please use
If no - let's use execute.

shubh220922 commented:
Work on this PR will resume in a few weeks upon conclusion of my exams.

shubh220922 commented:
I'm working on this pr now.

shubh220922 commented:
@romani @nrmancuso i have reverted back to the changes I had made by replacing ```verify()``` with ```execute()```. This does not seem to work though. I am trying to replace the ```verify()``` with ```verifyWithSuppressedParser()``` at the moment as my previous attempt failed. Will push the changes asap.

romani commented:
we just need to suppress PMD violation.
please rename test methods
testRemoteFileExternalResourceContentDoesNotChange
testLocalFileExternalResourceContentDoesNotChange
 to put 'UseCache' .... name should looks like  `testUseCache.....`
and let add xpath for suppression of PMD violation
at https://github.com/checkstyle/checkstyle/blob/master/config/pmd-test.xml#L124 to suppress this vioaltion on all methods of SuppressionFilterTest.java that starts with `testUseCache`.

shubh220922 commented:
@romani  I have made the changes you asked for. Facing slight diffculty in the PMD-test.xml file as the changes I made are incorrect. Could you please review what needs to be changed.

nrmancuso commented on config/pmd-test.xml:
```diff
@@ -140,7 +140,9 @@
                    | //ClassOrInterfaceDeclaration[@SimpleName='ArchUnitCyclesCheckTest']
-                       //MethodDeclaration"/>
+                       //MethodDeclaration
+                   | //ClassOrInterfaceDeclaration[@SimpleName='SuppressionFilterTest']
+                       //MethodDeclaration//Name[starts-with(@SimpleName, 'testUseCache')]"/>
```
@shubh220922 try to dump AST of the file to help you build/validate your XPath query: https://pmd.github.io/latest/pmd_devdocs_experimental_ast_dump.html

romani replied:
PMD has even UI tool to test xpath, please play with it.

romani replied:
try
`//MethodDeclaration[starts-with(@SimpleName, 'testUseCache')]"/>`

shubh220922 replied:
@romani 
build still failed after writing ```//MethodDeclaration[starts-with(@SimpleName, 'testUseCache')]"/>```
![image](https://user-images.githubusercontent.com/90137881/210445935-49902489-a169-4277-a51d-c7f52ec7bcf9.png)
Still showing the same error as before. 
```
[INFO] PMD Failure: com.puppycrawl.tools.checkstyle.filters.SuppressionFilterTest:163 Rule:JUnitTestsShouldIncludeAssert Priority:3 JUnit tests should include assert() or fail().
[INFO] PMD Failure: com.puppycrawl.tools.checkstyle.filters.SuppressionFilterTest:179 Rule:JUnitTestsShouldIncludeAssert Priority:3 JUnit tests should include assert() or fail().
```

shubh220922 replied:
@nrmancuso 

- [x] Downloaded the necessary softwares.
- [x] Dumped the AST of ```SuppressionFilterTest``` in an xml file.
- [ ] Attempted validating query multiple times but command seems to fail (CLI below):

```
PS C:\Users\shubh> xml --version
1.6.1
compiled against libxml2 2.9.1, linked with 20901
compiled against libxslt 1.1.28, linked with 10128
PS C:\Users\shubh> xml select -t -c "//ClassOrInterfaceDeclaration[@SimpleName='SuppressionFilterTest']//MethodDeclaration[starts-with(@SimpleName, 'testUseCache')]" "C:\Users\shubh\OneDrive\Desktop\checkstyle project\foo.xml"
PS C:\Users\shubh> xml select -t -c "//MethodDeclaration[starts-with(@SimpleName, 'testUseCache')]" "C:\Users\shubh\OneDrive\Desktop\checkstyle project\foo.xml"                                                                                                    
PS C:\Users\shubh> xml select -t -c "//ClassOrInterfaceDeclaration[@SimpleName='SuppressionFilterTest']//MethodDeclaration//Name[starts-with(@SimpleName, 'testUseCache')]" "C:\Users\shubh\OneDrive\Desktop\checkstyle project\foo.xml"
PS C:\Users\shubh> xml select -t -c "//MethodDeclaration//Name[starts-with(@SimpleName, 'testUseCache')]" "C:\Users\shubh\OneDrive\Desktop\checkstyle project\foo.xml"
```

Here expected was to get an output after execution of the commands above but was found no output. Am I querying the wrong way?

I am using xmlstarlet for query.

nrmancuso replied:
Please share the generated xml file by GitHub gist.

shubh220922 replied:
@nrmancuso here it is:

https://gist.githubusercontent.com/shubh220922/3db3a54bbc2edb9e77166274c4ee7ed1/raw/e757edaa045f4af2ada754f965feb2ae98915066/dump.xml

nrmancuso replied:
```
<MethodDeclaration Abstract='false' Arity='0' Default='false' Final='false' Image='' InterfaceMember='false' Kind='METHOD' MethodName='testUseCacheLocalFileExternalResourceContentDoesNotChange' Modifiers='1' Name='testUseCacheLocalFileExternalResourceContentDoesNotChange' Native='false' PackagePrivate='false' Private='false' Protected='false' Public='true' Static='false' Strictfp='false' Synchronized='false' SyntacticallyAbstract='false' SyntacticallyPublic='true' Transient='false' Void='true' Volatile='false'>
```
I do not see `SimpleName` attribute in this node, only `Name` and `MethodName`.

shubh220922 replied:
@nrmancuso below:

https://gist.githubusercontent.com/shubh220922/72138e2b956a0b7a49799c0c47df9379/raw/34ff3ad31a37ec28ac1887d875a23ea79731c1a5/query.xml

shubh220922 replied:
> I do not see `SimpleName` attribute in this node, only `Name` and `MethodName`.

Thanks for pointing out. 

now it is working.. i have shared the gist of the output @nrmancuso

nrmancuso replied:
Let's see if suppression works now  :)

shubh220922 replied:
@shubh220922 suppression seems successful. Maven = Success. Checking CI.

shubh220922 commented:
```run-inspections``` and ```sonarqube``` is failing. What are they pointing to? @nrmancuso

nrmancuso commented:
> `run-inspections` and `sonarqube` is failing. What are they pointing to? @nrmancuso
> 
> below: https://app.circleci.com/pipelines/github/checkstyle/checkstyle/16823/workflows/232ae426-848c-4906-b88b-9949ba5091ba/jobs/225540/steps?invite=true#step-103-66

Investigate and resolve errors, reports are generated for both of the tasks

romani commented:
@shubh220922

>There are inspection problems. Review results in 'ARTIFACTS' tab above.
Exited with code exit status 1

Please propose how to improve logging message above to be visible by users?

@nrmancuso , where this line comes from ?

shubh220922 commented:
@romani 
> Please propose how to improve logging message above to be visible by users?

I will let you know once I solve the CI to provide suggestions for possible changes to improve logging message.

shubh220922 commented:
@nrmancuso Even after trying multiple times, I can't get what the CI is asking me to fix. There is no simple explanation as to **what is wrong** and which file has the problem so that I can even think of fixing it. 

What am I supposed to fix?

nrmancuso commented:
Sonarqube: https://sonarcloud.io/project/issues?resolved=false&severities=BLOCKER&sinceLeakPeriod=true&pullRequest=12457&id=org.checkstyle%3Acheckstyle

A maintainer needs to suppress this after merge.

inspections failure: https://circleci-tasks-prod.s3.us-east-1.amazonaws.com/storage/artifacts/2233fdd7-79cb-481c-a64c-0df373c0d898/12a37e22-6271-4640-8949-db8d0d0c9c48/0/home/circleci/project/target/inspection-results/TestMethodWithoutAssertion.xml?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQVFQINEODJQDWTUT%2F20230104%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230104T203245Z&X-Amz-Expires=60&X-Amz-Security-Token=IQoJb3JpZ2luX2VjEK3%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJHMEUCIAD5ujrwkyU9ZH0wjl7I1koAxfGKwCO%2B3bik0nwjCMhuAiEAyEuL1fDTBayDpNR0pJ47BaMoeVkXo4SrSFdq0fcfVMIqtAII9f%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FARADGgwwNDU0NjY4MDY1NTYiDAFSCJWe8AWh2Yf4ayqIAmMvecVKX9Z82EQ5WpE136HbXMY3q1m1JXzTIlwQ%2FF6BbEC0IzmPWDDUPTRB0F1trVFGh829%2BLkggW7XpGBqxlaRHO94xdOO7fiUT%2FSixF3oeKxcgP3qfb5Dbuy6Z%2B%2BfZql%2BqOzSAy%2FGNC1503m74ZDC8WCpSr6NTVx8r2jaUA8WZ5zXh76BIkEAcCfj7dT3tPWQmEAYUOJ8pXNUciyHbo4srOnK66KHqkcfDYHrljvbWbd40Gx3bhptsoQ7KKF9y6FkBPeXWAJrA939l4aILELjQsk1oOkEcQZzKQZxjWhf1l9Q49DAxX8kZxhaKBRWQS2gaEaaj2hEM4yhMwyY3%2BqfcMNCYVzLUTCrvtedBjqdAd2Oss4iKCbppTVMawbyNV5IMIN2jgpLN5eXJQISou806%2FUL4rc9kXrZLWzszdYZ22v87EwUAy8MwgcGr6ZN8XHAcQAvMr4hyLIXmq9nCBIBPGnL%2FuZ3a1jCfa688hXisYesYprPZtCeOXIFvX7ehzD%2FMXasezQQlh0P1VvNA8ZLl4i9B8xdOZ0fntxcgVdp1kqTQguY0QXLVh92Dx4%3D&X-Amz-SignedHeaders=host&x-id=GetObject&X-Amz-Signature=034aa99ae1178611c98de6446f006297a5d39836a9978fef9c6faedf1846229e

You need to add a suppression in https://github.com/checkstyle/checkstyle/blob/229d55604e7ad604cda2d2235c9b1bd619977661/config/intellij-idea-inspections.xml#L4628

nrmancuso commented:
>@nrmancuso , where this line comes from ?

```
➜  checkstyle git:(master) ✗ ag -Q "There are inspection problems." --hidden         
.ci/idea_inspection.sh
52:    echo "There are inspection problems. Review results in 'ARTIFACTS' tab above."
55:    echo "There are inspection problems. Review results at $RESULTS_DIR folder. Files:"

```

shubh220922 commented:
Made some changes, checking CI here.

nrmancuso commented on config/intellij-idea-inspections.xml:
```diff
@@ -4638,7 +4638,8 @@
     ,org.checkstyle.base.AbstractItModuleTestSupport,verify.*, ,
-    ,com.tngtech.archunit.lang.ArchRule,check"/>
+    ,com.tngtech.archunit.lang.ArchRule,check ,
+    ,com.puppycrawl.tools.checkstyle.filters.SuppressionFilterTest, assert.*"/>
```
`package.name, methodnamePattern`

shubh220922 replied:
yes. I had figured that. pushing changes.

shubh220922 replied:
I have made some changes. Please check. CI is still showing error.

shubh220922 replied:
Done. Made some changes.

shubh220922 replied:
@nrmancuso ``run-inspections`` is failing. Seems like the suppression I created didn't work.

romani replied:
@shubh220922 , please download IDEA ,
follow https://checkstyle.org/idea.html#Inspections and play with config setup to make it be suppressed.

shubh220922 replied:
Done

shubh220922 commented:
@romani I worked with the config and inspections. Made some chages and added ```@SuppressWarnings("JUnitTestMethodWithNoAssertions")``` wherever required but, maven failed. So I reverted the changes. I have written the suppression in the pattern ``package.name, methodnamePattern`` in ``intellij-idea-inspections.xml`` but, CI is still failing.

romani commented on config/intellij-idea-inspections.xml:
```diff
@@ -4638,7 +4638,8 @@
     ,nl.jqno.equalsverifier.EqualsVerifier,verify.*, ,
     ,org.checkstyle.base.AbstractItModuleTestSupport,verify.*, ,
-    ,com.tngtech.archunit.lang.ArchRule,check"/>
+    ,com.tngtech.archunit.lang.ArchRule,check ,
```
`,com.tngtech.archunit.lang.ArchRule,check , ,`

two trailing commas are required.

shubh220922 replied:
Done.

romani reviewed (changes requested):
item:

shubh220922 commented:
I have worked with the config as you mentioned @romani, ```run-inspections```  is still failing.

romani commented on config/intellij-idea-inspections.xml:
```diff
@@ -4638,7 +4638,8 @@
     ,org.checkstyle.base.AbstractItModuleTestSupport,verify.*, ,
-    ,com.tngtech.archunit.lang.ArchRule,check"/>
+    ,com.tngtech.archunit.lang.ArchRule,check, ,
+    ,com.puppycrawl.tools.checkstyle.filters.SuppressionFilterTest,testUseCache.*"/>
```
should be:
`com.puppycrawl.tools.checkstyle.AbstractModuleTestSupport, execute`

as you show Inspection method that should be considered as "assert"

shubh220922 replied:
Done

romani reviewed (changes requested):
items:

romani reviewed (approved):
ok to merge
